src/shogun/regression/KernelRidgeRegression.cpp

```diff
@@ -43,86 +42,60 @@ void CKernelRidgeRegression::init()
+	SGMatrix<float64_t> kernel_matrix(kernel->get_kernel_matrix());
+	int32_t n = kernel_matrix.num_rows;
+	SGVector<float64_t> y = ((CRegressionLabels*)m_labels)->get_labels();
 
-	for(int32_t i=0; i < n; i++)
-		kernel_matrix.matrix[i+i*n]+=m_tau;
-
-	/* re-set alphas of kernel machine */
-	m_alpha=((CRegressionLabels*) m_labels)->get_labels_copy();
-
-	/* tell kernel machine that all alphas are needed as'support vectors' */
-	m_svs=SGVector<index_t>(m_alpha.vlen);
-	m_svs.range_fill();
-
-	if (get_alphas().vlen!=n)
+	for(index_t i=0; i<n; i++)
 	{
-		SG_ERROR("Number of labels does not match number of kernel"
-				" columns (num_labels=%d cols=%d\n", m_alpha.vlen, n);
+		alphas[i] = y[i];
+		kernel_matrix(i,i) += m_tau;
 	}
 
-	clapack_dposv(CblasRowMajor,CblasUpper, n, 1, kernel_matrix.matrix, n,
-			m_alpha.vector, n);
+	Map<MatrixXd> eigen_kernel_matrix(kernel_matrix.matrix, n, n);
+	Map<VectorXd> eigen_alphas(alphas.vector, n);
 
+	LLT<MatrixXd> llt;
+	llt.compute(eigen_kernel_matrix);
+	if (llt.info() != Eigen::Success)
+	{
+		SG_WARNING("Features covariance matrix was not positive definite\n");
+		return false;
+	}
+	eigen_alphas = llt.solve(eigen_alphas);
 	return true;
 }
 
-bool CKernelRidgeRegression::train_machine_gs()
+bool CKernelRidgeRegression::train_machine(CFeatures *data)
 {
-	int32_t n = kernel->get_num_vec_rhs();
-	int32_t m = kernel->get_num_vec_lhs();
-	ASSERT(m>0 && n>0)
+	check_and_set_alphas(data);
 
-	// re-set alphas of kernel machine
-	SGVector<float64_t> b;
-	float64_t alpha_old;
+	if(!solve_krr_system(m_alpha))
+		return false;
 
-	b=((CRegressionLabels*) m_labels)->get_labels_copy();
-	m_alpha=((CRegressionLabels*) m_labels)->get_labels_copy();
-	m_alpha.zero();
-
-	// tell kernel machine that all alphas are needed as 'support vectors'
-	m_svs=SGVector<index_t>(m_alpha.vlen);
+	/* tell kernel machine that all alphas are needed as'support vectors' */
+	m_svs = SGVector<index_t>(m_alpha.vlen);
 	m_svs.range_fill();
+	return true;
+}
 
-	if (get_alphas().vlen!=n)
-	{
-		SG_ERROR("Number of labels does not match number of kernel"
-				" columns (num_labels=%d cols=%d\n", m_alpha.vlen, n);
-	}
-
-	// Gauss-Seidel iterative method
-	float64_t sigma, err, d;
-	bool flag=true;
-	while(flag)
-	{
-		err=0.0;
-		for(int32_t i=0; i<n; i++)
-		{
-			sigma=b[i];
-			for(int32_t j=0; j<n; j++)
-				if (i!=j)
-					sigma-=kernel->kernel(j, i)*m_alpha[j];
-			alpha_old=m_alpha[i];
-			m_alpha[i]=sigma/(kernel->kernel(i, i)+m_tau);
-			d=fabs(alpha_old-m_alpha[i]);
-			if(d>err)
-				err=d;
-		}
-		if (err<=m_epsilon)
-			flag=false;
-	}
+bool CKernelRidgeRegression::load(FILE* srcfile)
+{
+	SG_SET_LOCALE_C;
+	SG_RESET_LOCALE;
+	return false;
+}
 
-	return true;
+bool CKernelRidgeRegression::save(FILE* dstfile)
+{
+	SG_SET_LOCALE_C;
+	SG_RESET_LOCALE;
+	return false;
 }
 
-bool CKernelRidgeRegression::train_machine(CFeatures *data)
+void CKernelRidgeRegression::check_and_set_alphas(CFeatures *data)
```

Comment:
I thought about this again.
Just allocate m_alphas in the train method before you call the solve method. The solve method assumes that memory is already allocated .... done